src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,475 @@
+  // multiple writes, it is recommended to resize the file to the expected
+  // size up front or to use the newWritable method to create a file with the
+  // expected size and add it into the target directory.
+  //
+  // Note that tryOpen doubles as both an existence check and a stat operation.
+  // When the createAs parameter is not specified, the method will return
+  // kj::none if the file or directory does not exist.
+  //
+  // If the path identifies a symbolic link, the symbolic link will be resolved
+  // and the target, if it exists, will be returned. If the target does not exist,
+  // kj::none will be returned.
+  virtual kj::Maybe<kj::OneOf<kj::Rc<File>, kj::Rc<Directory>>> tryOpen(
+      jsg::Lock& js, kj::PathPtr path, kj::Maybe<FsType> createAs = kj::none) = 0;
+
+  // Attempts to move the given file or directory into the path specified. The
+  // path must be relative to the current directory. If the path already exists
+  // an exception will be thrown. The name must not contain any path separators.
+  // If this directory is read only, an exception will be thrown.
+  virtual void add(jsg::Lock& js, kj::StringPtr name, Item entry) = 0;
+
+  struct RemoveOptions {
+    // If true and the node is a directory, remove all entries in the directory.
+    bool recursive = false;
+  };
+
+  // Tries to remove the file or directory at the given path. If the node does
+  // not exist, true will be returned. If the node is a directory and the recursive
+  // option is not set, an exception will be thrown if the directory is not empty.
+  // If the directory is read only, an exception will be thrown.
+  // If the node is a file, it will be removed regardless of the recursive option
+  // if the directory is not read only.
+  // The path must be relative to the current directory.
+  // Note that this method will only remove the node from this directory. If the
+  // node as references in other directories, those will not be removed.
+  // If the target is a symbolic link, the symbolic link will be removed but
+  // the target will not be removed.
+  virtual bool remove(
+      jsg::Lock& js, kj::PathPtr path, kj::Maybe<RemoveOptions> options = kj::none) = 0;
+
+  virtual kj::StringPtr jsgGetMemoryName() const = 0;
+  virtual size_t jsgGetMemorySelfSize() const = 0;
+  virtual void jsgGetMemoryInfo(jsg::MemoryTracker& tracker) const = 0;
+
+  // Creates a new readable/writable in-memory directory. This directory will
+  // not be initially included in a directory. To add it to a directory, use the
+  // add method. If the directory is not added, it will be deleted
+  // when the handle is dropped.
+  static kj::Rc<Directory> newWritable();
+
+  // Used to build a new read-only directory. All files and directories added
+  // may or may not be writable. The directory will not be initially included
+  // in a directory. To add it to a directory, use the add method.
+  class Builder {
```

Comment:
Is the only reason we actually need a Builder is because Directory doesn't support adding files with path separators?
e.g. tryOpen("dir_a/dir_b/file", mode=Create) is not supported and will not recursively create directories as required

Reply:
Builder is only used when constructing a read only directory. It's not used at all with directories that are already built.

Reply:
Can't we simply add a releaseAsReadOnly to a WritableDirectory then?

Reply:
Sure but I'm not convinced there's much benefit in doing so as it really wouldn't reduce the complexity at all.